contrib/database/sql/internal/dsn.go

```diff
@@ -135,3 +138,165 @@ func parseSQLServerDSN(dsn string) (map[string]string, error) {
 	delete(meta, "password")
 	return meta, nil
 }
+
+// parseSafe behaves like url.Parse, but if parsing fails it returns an
+// error with any credential part of the URL already scrubbed.
+func parseSafe(raw string) (*url.URL, error) {
+	u, err := url.Parse(raw)
+	if err == nil {
+		return u, nil
+	}
+	// url.Parse always wraps the real problem in *url.Error.
+	var ue *url.Error
+	if !errors.As(err, &ue) {
+		return nil, err
+	}
+	return nil, sanitizeURLError(ue)
+}
+
+// sanitizeURLError returns a copy of e whose URL field is redacted.
+func sanitizeURLError(e *url.Error) *url.Error {
+	// Best-case: we can still parse enough to use URL.Redacted().
+	if parsed, perr := url.Parse(e.URL); perr == nil {
+		e.URL = parsed.Redacted()
+		return e
+	}
+
+	// Fallback: use the comprehensive sanitize() function for all password patterns.
+	e.URL = sanitize(e.URL)
+	return e
+}
+
+// sanitizeError returns an error with sensitive information redacted.
+// It handles both URL errors and general errors containing DSN information.
+func sanitizeError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// Check if it's a URL error and use the existing sanitizer.
+	var ue *url.Error
+	if errors.As(err, &ue) {
+		return sanitizeURLError(ue)
+	}
+
+	// For other errors, sanitize the error message.
+	return fmt.Errorf("%s", sanitize(err.Error()))
+}
+
+func sanitize(msg string) string {
+	msg = sanitizeKeyValuePasswords(msg)
+	msg = sanitizeURLPasswords(msg)
+	msg = sanitizeMySQLPasswords(msg)
+	return msg
+}
+
+// Compiled regex patterns for password sanitization - compiled once at package init
+var (
+	keyValueSpacePattern = regexp.MustCompile(`(?i)(password|passwd|pwd)\s*=\s*(.+?)(\s+(?:host|user|port|database|dbname)\s*=)`)
+	keyValueSemiPattern  = regexp.MustCompile(`(?i)(password|passwd|pwd)\s*=\s*([^;]+)(;)`)
+	keyValuePattern      = regexp.MustCompile(`(?i)(password|passwd|pwd)\s*=\s*([^\s;]+)`)
+)
+
+// sanitizeKeyValuePasswords sanitizes password values in key=value format
+func sanitizeKeyValuePasswords(msg string) string {
+	result := keyValueSpacePattern.ReplaceAllString(msg, `$1=xxxxx$3`)
+	result = keyValueSemiPattern.ReplaceAllString(result, `$1=xxxxx$3`)
+	result = keyValuePattern.ReplaceAllString(result, `$1=xxxxx`)
+	return result
+}
+
+// sanitizeURLPasswords sanitizes passwords in URL format (user:pass@host).
+func sanitizeURLPasswords(msg string) string {
+	// Look for URL patterns and manually handle them to avoid issues with @ in passwords.
+	urlStart := strings.Index(msg, "://")
+	if urlStart == -1 {
+		return msg
+	}
```

Comment:
This is a fallback for an URL that did not successfully parse isn't it? Yet here we're failing open in cases where the URL is blatantly invalid?